### PR TITLE
SRTP/SRTCP KDF: add implementation

### PR DESCRIPTION
**Pull request by SparkiDev**
# Description

Add implementation of SRTP KDF and SRTCP KDF.
One shot APIs compatible with SP 800-135 and ACVP testing. Tests added to test.c.
Benchmarking added.
Doxygen added.

# Testing

./configure --enable-srtp-kdf
Tests added to test.c

# Checklist

 - [x] added tests
 - [x updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

**Comment by SparkiDev** [comment]
retest this please

**Comment by guidovranken** [comment]
Output keys remain (partially) uninitialized if too large:

```c
#include <wolfssl/options.h>
#include <wolfssl/wolfcrypt/kdf.h>

#define CF_CHECK_EQ(expr, res) if ( (expr) != (res) ) { goto end; }

int main(void)
{
    uint8_t key[32] = {0};
    uint8_t salt[10] = {0};
    uint8_t index[6] = {0};
#define KEYSIZE 256
    uint8_t* key1 = malloc(KEYSIZE);
    uint8_t* key2 = malloc(KEYSIZE);
    uint8_t* key3 = malloc(KEYSIZE);

    CF_CHECK_EQ(
            wc_SRTP_KDF(
                key, sizeof(key),
                salt, sizeof(salt),
                -1,
                index,
                key1, KEYSIZE,
                key2, KEYSIZE,
                key3, KEYSIZE), 0);

    {
        FILE* fp = fopen("/dev/null", "wb");
        fwrite(key1, KEYSIZE, 1, fp);
        fwrite(key2, KEYSIZE, 1, fp);
        fwrite(key3, KEYSIZE, 1, fp);
        fclose(fp);
    }

    free(key1);
    free(key2);
    free(key3);
end:
    return 0;
}
```

```
==1899383== Syscall param write(buf) points to uninitialised byte(s)
==1899383==    at 0x4B28077: write (write.c:26)
==1899383==    by 0x4AA8E8C: _IO_file_write@@GLIBC_2.2.5 (fileops.c:1181)
==1899383==    by 0x4AAA950: new_do_write (fileops.c:449)
==1899383==    by 0x4AAA950: _IO_new_do_write (fileops.c:426)
==1899383==    by 0x4AAA950: _IO_do_write@@GLIBC_2.2.5 (fileops.c:423)
==1899383==    by 0x4AA9F37: _IO_file_close_it@@GLIBC_2.2.5 (fileops.c:136)
==1899383==    by 0x4A9BF35: fclose@@GLIBC_2.2.5 (iofclose.c:53)
==1899383==    by 0x155575: main (in /home/jhg/cf-wolf-srtp-kdf/cryptofuzz/a.out)
==1899383==  Address 0x4c0fb10 is 0 bytes inside a block of size 4,096 alloc'd
==1899383==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==1899383==    by 0x4A9BD03: _IO_file_doallocate (filedoalloc.c:101)
==1899383==    by 0x4AABECF: _IO_doallocbuf (genops.c:347)
==1899383==    by 0x4AAAF2F: _IO_file_overflow@@GLIBC_2.2.5 (fileops.c:745)
==1899383==    by 0x4AA96B4: _IO_new_file_xsputn (fileops.c:1244)
==1899383==    by 0x4AA96B4: _IO_file_xsputn@@GLIBC_2.2.5 (fileops.c:1197)
==1899383==    by 0x4A9D3C0: fwrite (iofwrite.c:39)
==1899383==    by 0x155543: main (in /home/jhg/cf-wolf-srtp-kdf/cryptofuzz/a.out)
```

Would it be possible to return error in this case to prevent accidental misuse?

**Comment by SparkiDev** [comment]
Thanks Guido.

I used the wrong type for the keySz in the internal function.
This has been fixed now.

Let us know if there is anything else.

Sean :-)

**Comment by guidovranken** [comment]
Confirmed fixed, and I cannot find any other bugs with this code at the moment.

File: wolfcrypt/test/test.c [hunk 1]
```diff
@@ -24806,6 +24816,396 @@ WOLFSSL_TEST_SUBROUTINE wc_test_ret_t hpke_test(void)
...
+    word32 ksSz;
+} Srtp_Kdf_Tv;
+
+WOLFSSL_TEST_SUBROUTINE wc_test_ret_t srtpkdf_test(void)
+{
+    wc_test_ret_t ret = 0;
+    /* 128-bit key, kdrIdx = -1 */
+    WOLFSSL_SMALL_STACK_STATIC const byte key_0[] = {
+        0xc4, 0x80, 0x9f, 0x6d, 0x36, 0x98, 0x88, 0x72,
+        0x8e, 0x26, 0xad, 0xb5, 0x32, 0x12, 0x98, 0x90
```
Comment (JacobBarthelmeh) [inline wolfcrypt/test/test.c:24845]: is there a quick way to test interop of the implementation?

Comment (SparkiDev) [reply wolfcrypt/test/test.c:24845]: These tests vectors were taken from CAVP testing.

File: wolfcrypt/src/kdf.c [hunk 2]
```diff
@@ -870,4 +873,291 @@ int wc_SSH_KDF(byte hashId, byte keyId, byte* key, word32 keySz,
...
+ * @param [in]  indexSz  Size of index value in bytes.
+ * @param [out] block    First block to encrypt.
+ */
+static void wc_srtp_kdf_first_block(const byte* salt, word32 saltSz, int kdrIdx,
+        const byte* index, byte indexSz, unsigned char* block)
+{
+    word32 i;
+
+    /* XOR salt into zeroized buffer. */
+    for (i = 0; i < 14 - saltSz; i++)
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply wolfcrypt/src/kdf.c]: Replaced with WC_SRTP_MAX_SALT.

File: wolfcrypt/src/kdf.c [hunk 3]
```diff
@@ -870,4 +873,291 @@ int wc_SSH_KDF(byte hashId, byte keyId, byte* key, word32 keySz,
...
+    int ret = 0;
+    byte block[AES_BLOCK_SIZE];
+#ifdef WOLFSSL_SMALL_STACK
+    Aes* aes = NULL;
+#else
+    Aes aes[1];
+#endif
+
+    /* Validate parameters. */
+    if ((key == NULL) || (keySz > 32) || (salt == NULL) || (saltSz > 14) ||
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply wolfcrypt/src/kdf.c]: Replaced with AES_256_KEY_SIZE

**Review by JacobBarthelmeh — APPROVED** [review]

**Pull request merged**